iron-input.html

```diff
@@ -125,13 +184,49 @@
      * @suppress {checkTypes}
      */
     _bindValueChanged: function() {
-      if (this.value !== this.bindValue) {
-        this.value = !(this.bindValue || this.bindValue === 0) ? '' : this.bindValue;
+      var newValue = !(this.bindValue || this.bindValue === 0) ? '' : this.bindValue;
+      if (this.value !== newValue) {
+        this.value = newValue;
+        this._updateBindValues();
       }
       // manually notify because we don't want to notify until after setting value
       this.fire('bind-value-changed', {value: this.bindValue});
     },
 
+    _bindValueAsNumberChanged: function() {
+      var newValue = this.bindValueAsNumber;
+      if ((this.type === 'number' || this.type === 'date') && this.valueAsNumber !== newValue) {
+        try {
+          this.valueAsNumber = newValue;
+        } catch (e) {
+          console.warn('bindValueAsNumber: could not set ' + newValue);
+        }
+        this._updateBindValues();
+      }
+    },
+
+    _bindValueAsDateChanged: function() {
+      var oldValue = this._toDateString(this.valueAsDate);
+      var newValue = this._toDateString(this.bindValueAsDate);
+      if (this.type === 'date' && oldValue !== newValue) {
+        this.valueAsDate = newValue ? new Date(newValue) : null;
+        this._updateBindValues();
+      }
+    },
+
+    _toDateString: function(date) {
+      var d;
+      if (date instanceof Date) {
```

Comment:
Hmmmm I'm not sure I understand why is all of this needed? Like, isn't `Date.toDateString()` a thing? Also, when you're calling it above, shouldn't you jsut save `bindValueAsDate` to whatever `valueAsDate` returns, since that will already be in the right format?